### ISSUE TITLE
[transpileDeclaration API][5.5] More missing type-only imports

### ISSUE THREAD

**Issue by MichaelMitchell-at** [microsoft/TypeScript#58511]
### 🔎 Search Terms

transpiledeclaration api emit missing import

### 🕗 Version & Regression Information

- This changed in commit or PR https://github.com/microsoft/TypeScript/pull/58261 @weswigham 

### ⏯ Playground Link

https://www.typescriptlang.org/dev/bug-workbench/?target=7&ts=5.5.0-dev.20240512#code/PTAEAEGcBcCcEsDG0Bco4FcCmAoEEATLRAGwENYzp4B7AOzU13ygAsaB3AUQFt5o8YNp179oWAgDF4JLGjIA6AguiRBESO259o4qTLmgARkpVr14AGYG6ZHocWqcWAB4AHGrGih4dcbEsyRCxQAEk6UABvHFBQAG0AaywATzQYBDoAcwBdNAw6BLpOOgBuHABfZ3dPb19-QODQAHkMb2jYxJS0uF8cvIKijlKKqo8vdGS3EIAhMkgQgF5QAHIyZdAAHxWjZbKLa1lbezQTJ3geMbaJqbC6ABprkJboB+hJmbmQ8tBLWBoeUAAIgUwDIgL2rkuoEQ9BgoBoRgAVsRvEt2qACDQACqsXoAHixoFc4joBEgoFm8wAfAAKAD6qFAWIedN8aHCcSx2QAlGhnpzslEYrFQLAsNAMLBhrFyncKmUgA

### 💻 Code

```ts
// @strict: true
// @declaration: true
// @showEmit
// @showEmittedFile: a.d.ts
// @showEmittedFile: b.d.ts

// @filename: a.ts
export interface In {
  [key: string]: unknown;
}
export interface Out {
  [key: string]: unknown;
}
export type Base = 'a' | 'b';

// @filename: b.ts
import { type In, type Out, type Base } from "./a";

export const object = {
  doThing<T extends Base>(_t: T, _in: In[T]): Out[T] {
    return;
  },
};
```

### 🙁 Actual behavior

When using the API, `b.d.ts` is emitted as
```ts
import { type In } from "./a";
export declare const object: {
    doThing<T extends Base>(_t: T, _in: In[T]): Out;
};
```
while when using `tsc` it gets emitted as
```ts
import { type In, type Out, type Base } from "./a";
export declare const object: {
    doThing<T extends Base>(_t: T, _in: In[T]): Out[T];
};
```
(interestingly the Debug tab of the workbench link shows something that's partially wrong)

### 🙂 Expected behavior

Using the API should emit
```ts
import { type In, type Out, type Base } from "./a";
export declare const object: {
    doThing<T extends Base>(_t: T, _in: In[T]): Out[T];
};
```

**Comment by MichaelMitchell-at** [comment]
I'm somewhat worried that trying to adapt checker.ts to also be able to purely emit declarations will lead to more problems like these to continue to arise. It seems harder to reason about than the behavior we'd expect with isolated declarations - a mostly syntactic transformation. I'm concerned that these discrepancies will lead to compatibility issues when migrating to alternative implementations such as likely ones like [ESBuild](https://github.com/evanw/esbuild/issues/95#issuecomment-1559710310).

**Comment by weswigham** [comment]
This is partially fixed by #58522 - the `Out` is elided because we don't traverse the `Out[T]` node for reuse (and instead use an `Out`-printing error type node). The `Base` reference is a separate issue, so this gets to stay open - looks like we're failing to collect references from the type parameter constraint position in our references visitor (which is separate from the checker's node builder and largely standalone - the references visitor was always intended to be reused since import tracking is common functionality that all declaration emitters need to support elision).
